presto-main/src/main/java/io/prestosql/operator/HttpPageBufferClient.java

```diff
@@ -414,21 +425,40 @@ public void onFailure(Throwable t)
                 checkNotHoldsLock(this);
 
                 Throwable throwable = rewriteException(t);
-                if (!(throwable instanceof PrestoException) && backoff.failure()) {
-                    String message = format("%s (%s - %s failures, failure duration %s, total failed request time %s)",
-                            WORKER_NODE_ERROR,
-                            uri,
-                            backoff.getFailureCount(),
-                            backoff.getFailureDuration().convertTo(SECONDS),
-                            backoff.getFailureRequestTimeTotal().convertTo(SECONDS));
-                    if (querySnapshotManager != null) {
-                        // Snapshot: recover from remote server errors
-                        log.debug(throwable, "Snapshot: remote task failed with resumable error: %s", message);
-                        querySnapshotManager.cancelToResume();
-                        handleFailure(throwable, resultFuture);
-                        return;
+                if (!(throwable instanceof PrestoException)) {
+                    boolean hasFailed;
+                    if (timeoutFailureDetection) {
```

Comment:
timeoutBased Failure or Backoff retry all these should be abstracted from here..
can expose interface from FailureDetector like **failureDetector.hasFailed()**  inside it accordingly states can be checked and failure decided.